torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -173,6 +222,11 @@ void DistEngine::computeDependencies(
     for (const auto& recvBackwardEdge : recvBackwardEdges) {
       graphTask->exec_info_[recvBackwardEdge.function.get()].needed_ = true;
     }
+    // Mark all output nodes as needed.
+    for (const auto& outputEdge : outputEdges) {
+      const auto func = outputEdge.function.get();
+      graphTask->exec_info_[func].needed_ = true;
```

Comment:
All the exec_info_ can be removed now. Because you want the whole graph to be processed.

Reply:
Will do that.

@mrshenli @pritamdamania  could you also help confirm that `exec_info_` can be deleted?

Reply:
@albanD I think I misunderstood you. What does "exec_info_ can be removed" mean? `exec_info_` is a field of `GraphTask` and is also used by the local engine to determine whether to apply a function.

Reply:
You can double check [this comment in the code](https://github.com/pytorch/pytorch/blob/20b18a58f19a9b5907697e4b1ffe075d3fc2ec52/torch/csrc/autograd/engine.h#L71-L77). But the exec_info_ field has a special semantic when its left empty.

Namely:
- exec_info_ == nullptr corresponds to a `loss.backward()` in python: the whole graph is processed with no exception.
- exec_info_ != nullptr corresponds to `autograd.grad(loss, inputs)` in python: only the graph between loss and inputs is processed.

Now that you change the graph to remove the AccumulateGrad, you actually want to run the engine just like `.backward()` as you want everything to run.

Reply:
How can `exec_info_` be `nullptr`? It's of type `std:: unordered_map` instead of a pointer. I guess you meant the map is empty.

I'm not sure we want everything to run. The signature of distributed backward is `backward(context_id: int, roots: List[Tensor], retain_graph = False)`. Only nodes of the autograd graph reachable from `roots` need to be run.
@mrshenli @pritamdamania87

Reply:
We always start at the roots, so everything that is not reachable from the roots won't run.
The signature is the same as the regular engine backward(). Which maps to an empty exec_info_.

Yes sorry, I meant `.empty()` not nullptr, you can see the check [here](https://github.com/pytorch/pytorch/blob/cd9d9a2235567eb76743a5d863e0f2d4074d8956/torch/csrc/autograd/engine.cpp#L559).

Reply:
Yes sorry I meant empty, not nullptr.

The backward is the same as the regular engine's backward that runs without exec_info_.
The current implementation uses exec_info_ because it needs to bypass the `AccumulateGrad` Nodes that should not run. In this case, since you remove them, this is not needed anymore.